File: MazePrograms/room.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <cstdint>

class Room
{
	std::uint8_t dimensionX, dimensionY, maxConnectedRooms, numConnectedRooms;
	std::vector<Room*> connectedRooms; //pointers to the rooms connected to this room
	public:
	Room() //http://www.cplusplus.com/reference/random/uniform_int_distribution/uniform_int_distribution/
	{
		// unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
		// std::default_random_engine generator (seed);
		// std::uniform_int_distribution<int> dimensionGenerator (1,5);
		// dimensionX = dimensionGenerator(generator);
		// dimensionY = dimensionGenerator(generator);
		dimensionX = 1;
		dimensionY = 1;
		maxConnectedRooms = dimensionX*2 + dimensionY*2;
		numConnectedRooms = 0;
	}
	~Room()
	{}

	void connectRooms(Room &r)
	{
		std::cout << "Worked???" << std::endl;
		(*connectedRooms.end()) = &r;
	}
	
	std::uint8_t getNumConnectedRooms()
	{
		return numConnectedRooms;
	}

	std::uint8_t getMaxConnectedRooms()
	{
		return maxConnectedRooms;
	}
};

int main(int argc, char const *argv[])
{
	using std::cout; using std::endl;
	std::uint8_t mazeX = 5, mazeY = 5;
	std::vector<std::vector<Room> > maze (mazeX,std::vector<Room> (mazeY));

	for (auto &vec : maze)
	{
		for (auto &room : vec)
		{
			room = Room();
		}
	}

	// for (auto &v : maze)
	// {
	// 	static int i = 0;
	// 	for (auto &r : v)
	// 	{
	// 		cout << (int)r.getNumConnectedRooms() << " " << i++ << endl;
	// 	}
	// }

	unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
	std::default_random_engine generator (seed);
	std::uniform_int_distribution<int> chooseX (0,mazeX-1);
	std::uniform_int_distribution<int> chooseY (0,mazeY-1);

	std::uint8_t startX = chooseX(generator), startY = chooseY(generator);

	Room *start = &maze[startX][startY]; //start is a pointer to the chosen starting room

	cout << (int)startX << " " << (int)startY << endl << static_cast<int>((*start).getMaxConnectedRooms()) << endl;

	std::uniform_int_distribution<int> chooseNeighbor (0,(*start).getMaxConnectedRooms());
	switch(chooseNeighbor(generator))
	{
		case 0:
			(*start).connectRooms(maze[startX][startY-1]);
			maze[startX][startY-1].connectRooms((*start));
		// case 1:
		// 	(*start).connectRooms(maze[startX+1][startY]);
		// 	maze[startX+1][startY].connectRooms((*start));
		// case 2:
		// 	(*start).connectRooms(maze[startX][startY+1]);
		// 	maze[startX][startY+1].connectRooms((*start));
		// case 3:
		// 	(*start).connectRooms(maze[startX-1][startY]);
		// 	maze[startX-1][startY].connectRooms((*start));
		default:
			cout << "Error: invalid neighbor code" << endl;
	}

	return 0;
}

File: MazePrograms/eclipse_workspace/MazeGenerationSearch/src/Room.java
public class Room
{
	private int dimensionX, dimensionY, maxConnectedRooms, numConnectedRooms;
	private Room[] connectedRooms; //pointers to the rooms connected to this room in the order n,e,s,w

	public Room()
	{
		dimensionX = 1;
		dimensionY = 1;
		maxConnectedRooms = dimensionX*2 + dimensionY*2;
		numConnectedRooms = 0;
		connectedRooms = new Room[maxConnectedRooms];
	}


	/*
	 * returns the integer direction of the new room from the given room
	 * 0 - North
	 * 1 - East
	 * 2 - South
	 * 3 - West
	 * -1 - connection failed because there are no available spaces
	 */
	public int connectRooms(Room r)
	{
		if(numConnectedRooms == maxConnectedRooms)
			return -1;
		connectedRooms[numConnectedRooms] = r;
		return numConnectedRooms;
	}

	
	/*
	 * returns the integer direction of the new room from the given room
	 * 0 - North
	 * 1 - East
	 * 2 - South
	 * 3 - West
	 * -1 - connection failed because there are no available spaces or that space was taken
	 */
	public int connectRooms(Room r, int direction)
	{
		if(numConnectedRooms == maxConnectedRooms || connectedRooms[direction] != null)
			return -1;
		connectedRooms[direction] = r;
		numConnectedRooms++;
		return direction;
	}
	
	public int getNumConnectedRooms()
	{
		return numConnectedRooms;
	}

	public int getMaxConnectedRooms()
	{
		return maxConnectedRooms;
	}
	
	public String[] toStringArray()
	{
		String[] roomString = new String[3];
		roomString[0] = (connectedRooms[0] == null) ? "#---#" : "#   #";
		roomString[1] = (connectedRooms[3] == null) ? "|   " : "    ";
		roomString[1] += (connectedRooms[1] == null) ? "|" : " ";
		roomString[2] = (connectedRooms[2] == null) ? "#---#" : "#   #";		
		return roomString;
	}
	
	public String[] toStringArray(int roomType) //0 if start, 1 if end, -1 if neither
	{
		String room = " ";
		if (roomType == 0)
			room = "S";
		else if (roomType == 1)
			room = "E";
		String[] roomString = new String[3];
		roomString[0] = (connectedRooms[0] == null) ? "#---#" : "#   #";
		roomString[1] = (connectedRooms[3] == null) ? "| " + room + " " : "  " + room + " ";
		roomString[1] += (connectedRooms[1] == null) ? "|" : " ";
		roomString[2] = (connectedRooms[2] == null) ? "#---#" : "#   #";		
		return roomString;
	}
	
//	public static void main(String[] args)
//	{
//		int mazeX = 5, mazeY = 5;
//		Room[][] maze = new Room[mazeX][mazeY];
//
//		for (int i = 0; i < maze.length; i++)
//		{
//			for (int j = 0; j < maze[i].length; j++)
//			{
//				maze[i][j] = new Room();
//			}
//		}
//
//		int startX = (int)(Math.random()*(mazeX-1)),
//			startY = (int)(Math.random()*(mazeY-1));
//
//		Room start = maze[startX][startY]; //start is the chosen starting room
//
//		switch((int)(Math.random()*(start.getMaxConnectedRooms()-1))) //FIX ARRAYINDEXOUTOFBOUNDS
//		{
//		case 0:
//			start.connectRooms(maze[startX][startY-1]);
//			maze[startX][startY-1].connectRooms(start);
//			break;
//		case 1:
//			start.connectRooms(maze[startX+1][startY]);
//			maze[startX+1][startY].connectRooms(start);
//			break;
//		case 2:
//			start.connectRooms(maze[startX][startY+1]);
//			maze[startX][startY+1].connectRooms(start);
//			break;
//		case 3:
//			start.connectRooms(maze[startX-1][startY]);
//			maze[startX-1][startY].connectRooms(start);
//			break;
//		default:
//			System.out.println("Error: invalid neighbor code");
//		}
//	}
}